crates/bevy_render/src/lib.rs

```diff
@@ -330,9 +315,36 @@ impl Plugin for RenderPlugin {
     }
 }
 
-/// Applies the
+/// A "scratch" world used to avoid allocating new worlds every frame when
+/// swapping out the [`MainWorld`] for [`RenderStage::Extract`].
+#[derive(Resource, Default)]
+struct ScratchMainWorld(World);
+
+/// Executes the [`Extract`](RenderStage::Extract) stage of the renderer.
+/// This updates the render world with the extracted ECS data of the current frame.
+fn extract(main_world: &mut World, render_app: &mut App) {
```

Comment:
I moved extract back into a function to try and keep the diff down for the final pr.